### ISSUE TITLE
Memory leak in ClpSimplex, Osi 0.97.0 rev. 193

### ISSUE THREAD

**Issue by svigerske** [coin-or/Osi#52]
Issue created by migration from Trac.

Original creator: stoinski

Original creation time: 2008-04-01 14:10:38

Assignee: @mjsaltzman

Version: 0.96

Keywords: ClpPresolve ClpSimplex presolvedModel memory leak

Dear COIN,
ClpSimplex::initialSolve() creates instances of ClpSimplex using model2 = pinfo.presolvedModel() (ClpPresolve::presolvedModel()) which are not deleted under circumstances.
I solved the problem for my purposes by deleting the instance pointed by ClpPresolve::presolvedModel_ in the destructor of ClpPresolve. In order to avoid segmentation faults in certain cases, after ClpSimplex::initialSolve() still decided to delete the instance pointed by model2, the deletion was prohibited in ClpSimples::initialSolve().
I attach a diff file for my solution. Please see it as a proposal and a complement to the bug description rather than a reliable solution. After patching with this diff, the cleanup works correctly for at least my optimization problems.
I didn't investigate the bug for the recent version 0.98 of Osi. A simple test basing on counting of new and delete calls for my problem shows memory leaks for this new version as well.
Please let me know, if you need any test data for the problem detection. I fill the model using directly the C++ interface, but I hope, I'll be able to generate an MPS file for it.
Regards and thanks in advance
Daniel

**Comment by svigerske** [comment]
Attachment [ClpSimplex_leak.diff](https://www.coin-or.org/issue-attachments/Osi/ticket52/ClpSimplex_leak.diff) by stoinski created at 2008-04-01 14:11:37

**Comment by svigerske** [comment]
Attachment [basic.cpp](https://www.coin-or.org/issue-attachments/Osi/ticket52/basic.cpp) by stoinski created at 2008-04-01 15:16:40

**Comment by svigerske** [comment]
Comment by stoinski created at 2008-04-01 15:25:17

I generated an MPS file for my problem using OsiClpSolverInterface::writeMps() and runned this file using both Osi-0.97 and Osi-0.98 with a slightly modified program basic.cpp found in Osi/examples/basic.cpp. I attached the modified version, which primitively traces new and delete calls. Comparing the output of this program for new and delete, also looking in /proc/self/stat, tasklist.exe or top shows, that the memory usage is growing and not all resources are released correctly.
Unfortunally my mps file is quite huge (24MB, 1.6MB compressed) and I can't upload it to this wiki page. But I can send it to you any time, if you ask.
Regards
Daniel
